handlers/src/lib.rs

```diff
@@ -0,0 +1,99 @@
+#[macro_use]
+extern crate serde_derive;
+
+use actix_web::FromRequest;
+use actix_web::{dev::Handler, HttpRequest, Path, Query};
+use git::{GitHelper, GitOps};
+use std::path::PathBuf;
+
+#[derive(Deserialize)]
+pub struct PathParams {
+    repo: String,
+}
+
+#[derive(Deserialize)]
+pub struct QueryParams {
+    reference: String,
+    file: String,
+}
+
+pub struct RepoHandler {
+    pub repo_root: String,
+    git_ops: Box<GitOps>,
+}
+
+impl<S> Handler<S> for RepoHandler {
+    type Result = String;
+
+    fn handle(&self, req: &HttpRequest<S>) -> Self::Result {
+        let path_params = Path::<PathParams>::extract(req).unwrap();
+        let query_params = Query::<QueryParams>::extract(req).unwrap();
+        let repo_path: PathBuf = [&self.repo_root, &path_params.repo].iter().collect();
```

Comment:
I think this is wildly unsafe — there's nothing stopping me from pointing it at `/etc/passwd` or some actual repo somewhere else on your system and play about until I get an error message or response to contain something interesting.

The way to fix this is to have a config that goes `name -> path` and require the param to mention the name, looking up the path server side. That is, effectively, whitelist the repos this tool can point to.

Perhaps the `toml` crate (which provides deserialisation of a toml file using serde) works? https://crates.io/crates/toml

Reply:
I do like the idea. For the sake of moving forward (and given that this is not yet production ready code) I propose to create an issue to have a list of allowed repos. This could be automatically populated from the `repo_root` path. And we could also expose an endpoint that returns the list of repos configured.